### DISCUSSION TITLE
draw_nine_patch

### DISCUSSION THREAD

**Discussion by Shadowblitz16** [godotengine/godot-proposals#6746] [2D]
### Describe the project you are working on

I am trying to extend the control node into a widget class that suits my needs better
However I need to be able to draw a nine patch

### Describe the problem or limitation you are having in your project

I don't exactly know how to draw a nine patch and a draw function/method would be nice to have

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

It allows people to quickly draw nine patches from any canvas node

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Idk what the api will looks like but perferablity it would take a rectangle and patch margin and a x and y stretch mode 

### If this enhancement will not be used often, can it be worked around with a few lines of script?

It would be used in cases built in nodes do not cut it and we need to draw a nine patch

### Is there a reason why this should be core and not an add-on in the asset library?

drawing functions are already part of the engine this would just be a addition

**Comment by Calinou** [comment]
This already exists as `RenderingServer.canvas_item_add_nine_patch()`. This is what NinePatchRect internally calls: https://github.com/godotengine/godot/blob/24cb43a8741c7b10abbbbc77bb6e2bc188662ce0/scene/gui/nine_patch_rect.cpp#L49

This could probably be exposed as a high-level function in CanvasItem, like `CanvasItem.draw_multimesh()`.

In 3.x, use VisualServer instead of RenderingServer.

**Reply by Shadowblitz16** [reply]
Wait does this add it or just render it for one frame?

**Reply by KoBeWi** [reply]
It draws it, just like the drawing functions of CanvasItem.

**Reply by Shadowblitz16** [reply]
so why is there a api in consistency? is it possible to move or copy it to a Node2D and Control draw function?

**Reply by KoBeWi** [reply]
> so why is there a api in consistency?

It's not inconsistency. There are methods like `canvas_item_add_circle()` or `canvas_item_add_polygon()`, which are called internally by CanvasItem. The ninepatch one just wasn't added to CanvasItem and only exists in RenderingServer.

> is it possible to move or copy it to a Node2D and Control draw function?

Yes. Not sure why it wasn't done.

**Reply by dalexeev** [reply]
* See godotengine/godot#76397.
